src/librustc_typeck/check/mod.rs

```diff
@@ -1142,7 +1142,9 @@ fn check_specialization_validity<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     if let Some(parent) = parent {
         if parent.item.is_final() {
-            report_forbidden_specialization(tcx, impl_item, parent.node.def_id());
+            if !tcx.impl_is_default(parent.node.def_id()) {
+                report_forbidden_specialization(tcx, impl_item, parent.node.def_id());
+            }
```

Comment:
@nagisa 
When i first start to do this PR I was thinking about some propagation during lowering but I choose to not do it since I should have done something on [hir/print.rs](https://github.com/rust-lang/rust/pull/37860/files/715811d0be05dcdc55b44f97d9fd2cd1eb7eee05#diff-c61bca06497a95ede4372cc2ad94038c) in order to show the defaultness info before the lowering occurred. An `is_final` method should be placed on the `NodeItem struct`, somthing like the following

```
impl<'a, 'gcx, 'tcx> NodeItem<hir::Defaultness> {
    pub fn is_final(self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
        self.item.is_final() || !tcx.impl_is_default(self.node.def_id())
    }
}
```
Then in typecheck just call `parent.is_final()`

Do you think it could be ok?